### ISSUE TITLE
Parse llvm bitcode in natively in rust

### ISSUE THREAD

**Issue by jrmuizel** [cdisselkoen/llvm-ir#1]
This would allow dropping the llvm dependency and avoid the problem of things not being reflected in the c bindings.

**Comment by cdisselkoen** [comment]
Hi,

I somehow completely missed this issue when it was opened - my apologies for responding almost two months later.  This is an excellent issue and eventual goal, which I'm happy to support.  I personally don't have plans in the near term to spend the (probably large) amount of effort to accomplish this; the current system relying on `llvm-sys` is "good enough" for most purposes.  That said, I'm happy to take contributions towards this end.

An alternate way that the limitations of the c bindings could potentially be circumvented would be for us to create our own C++ library with a C interface which exposes getters for the missing items.  Of course, at that point we might as well just upstream the additional getters to the LLVM C API, which I believe the LLVM people are happy to take contributions for.

Thanks for your interest in `llvm-ir`!  I'll leave this issue open to keep conversation going about this topic.

**Comment by jrmuizel** [comment]
It looks like https://github.com/woodruffw/mollusc gets part of the way there with its `llvm-bitstream` crate.

**Comment by cdisselkoen** [comment]
Hey this is cool!  thanks for the pointer!

**Comment by woodruffw** [comment]
Yep! The `llvm-bitcursor` and `llvm-bitstream` crates in that repo are pretty low-level, but should be pretty stable in terms of functionality (and LLVM changes very little at that layer). 

I haven't worked much on the "mapping" layer above those crates in the past few months, though.
